pallets/treasury/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+};
+use frame_system::{self as system, ensure_root, ensure_signed};
+use sp_runtime::traits::Saturating;
+use sp_std::{convert::TryFrom, prelude::*};
+
+pub type ProposalIndex = u32;
+
+type BalanceOf<T> = <T as CommonTrait>::Balance;
+type Identity<T> = identity::Module<T>;
+
+pub trait Trait: frame_system::Trait + CommonTrait + BalancesTrait + IdentityTrait {
+    // The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+}
+
+pub trait TreasuryTrait<Balance> {
+    fn disbursement(target: IdentityId, amount: Balance);
+    fn balance() -> Balance;
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Treasury {
+        pub Balance get(fn balance) config(): BalanceOf<T>;
+    }
+}
+
+decl_event!(
+    pub enum Event<T> where
+    <T as CommonTrait>::Balance
+    {
+        /// Disbursement to a target Identity.
+        /// (target identity, amount)
+        TreasuryDisbursement(IdentityId, Balance),
+
+        /// Treasury reimbursement.
+        TreasuryReimbursement(Balance),
+    }
+);
+
+decl_error! {
+    /// Error for the treasury module.
+    pub enum Error for Module<T: Trait> {
+        /// Proposer's balance is too low.
+        InsufficientBalance,
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// It transfers balances from treasury to each of beneficiaries and the specific amount
+        /// for each of them.
+        ///
+        /// # Error
+        /// * `BadOrigin`: Only root can execute transaction.
+        /// * `InsufficientBalance`: If treasury balances is not enough to cover all beneficiaries.
+        pub fn disbursement(origin, beneficiaries: Vec< Beneficiary<T::Balance>>) -> DispatchResult
+        {
+            ensure_root(origin)?;
+
+            // Ensure treasury has enough balance.
+            let total_amount = beneficiaries.iter().fold( 0.into(), |acc,b| b.amount.saturating_add(acc));
+            ensure!(
+                Self::balance() >= total_amount,
+                Error::<T>::InsufficientBalance);
+
+            beneficiaries.into_iter().for_each( |b| {
+                Self::unsafe_disbursement(b.id, b.amount);
+                Self::deposit_event(RawEvent::TreasuryDisbursement(b.id, b.amount));
+            });
+            Ok(())
+        }
+
+        /// It transfers the specific `amount` from `origin` account into treasury.
+        ///
+        /// Only accounts which are associated to an identity can make a donation to treasury.
+        pub fn reimbursement(origin, amount: T::Balance) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let sender_key = AccountKey::try_from(sender.encode())?;
+            let _did = Context::current_identity_or::<Identity<T>>(&sender_key)?;
+
+            let _ = balances::Module::<T>::withdraw(
```

Comment:
Since the imbalance is not being handled here, the system will automatically reduce total supply when this goes out of scope. Either we need to manually offset the imbalance to prevent total supply from changing or even better, we should store treasury balance in the balances module itself rather than in this module. You can use `<Module<T>>::account_id()` to get account id unique to this module and store its balance in the balances module/`T::Currency` using that.

Reply:
I see you have implemented the `onUnbalanced` trait but I think you might need to manually call it via `on_unbalanced` and still manage the drop behavior.

Reply:
`onUnbalanced` trait is automatically called during the split of the fees (80/20). See `SplitToWays` at `runtime.rs` file.